File: python/uverify_sdk/models/certificate.py
"""Certificate-related data models."""
from __future__ import annotations

import json as _json
from dataclasses import dataclass
from typing import Dict, Optional, Union


@dataclass
class CertificateData:
    """
    Certificate data provided when building a new certification transaction.

    Attributes:
        hash:      SHA-256 or SHA-512 hex hash of the data to certify.
        metadata:  Optional metadata to attach — either a pre-serialised JSON
                   string or a plain ``dict`` (automatically serialised).
        algorithm: Hashing algorithm, e.g. "SHA-256" or "SHA-512".
    """

    hash: str
    metadata: Optional[Union[str, Dict[str, str]]] = None
    algorithm: Optional[str] = None

    def to_dict(self) -> dict:
        d: dict = {"hash": self.hash}
        if self.metadata is not None:
            d["metadata"] = (
                _json.dumps(self.metadata)
                if isinstance(self.metadata, dict)
                else self.metadata
            )
        if self.algorithm is not None:
            d["algorithm"] = self.algorithm
        return d


@dataclass
class CertificateResponse:
    """
    A single certificate as returned by the verification endpoints.

    Attributes:
        hash:             SHA-256 or SHA-512 hash of the certified data.
        address:          Cardano address that issued the certificate.
        block_hash:       Hash of the containing block.
        block_number:     Block height.
        transaction_hash: Cardano transaction hash.
        slot:             Slot number.
        creation_time:    ISO-8601 timestamp.
        metadata:         Optional attached metadata.
        issuer:           Optional issuer identifier.
    """

    hash: str
    address: str
    block_hash: str
    block_number: int
    transaction_hash: str
    slot: int
    creation_time: str
    metadata: Optional[str] = None
    issuer: Optional[str] = None

    @classmethod
    def from_dict(cls, data: dict) -> "CertificateResponse":
        return cls(
            hash=data["hash"],
            address=data["address"],
            block_hash=data["blockHash"],
            block_number=data["blockNumber"],
            transaction_hash=data["transactionHash"],
            slot=data["slot"],
            creation_time=data["creationTime"],
            metadata=data.get("metadata"),
            issuer=data.get("issuer"),
        )

File: typescript/src/types/certificate.ts
/**
 * A single certificate entry as stored on-chain and returned by the verification endpoints.
 */
export interface CertificateResponse {
  /** SHA-256 or SHA-512 hash of the certified data. */
  hash: string;
  /** Cardano address that issued the certificate. */
  address: string;
  /** Hash of the block that contains this certificate. */
  blockHash: string;
  /** Block height at which the certificate was recorded. */
  blockNumber: number;
  /** Transaction hash on the Cardano blockchain. */
  transactionHash: string;
  /** Slot number at which the certificate was recorded. */
  slot: number;
  /** ISO-8601 timestamp of the block. */
  creationTime: string;
  /** Optional metadata attached to the certificate. */
  metadata?: string;
  /** Optional issuer identifier. */
  issuer?: string;
}

/**
 * Certificate data provided when building a new certification transaction.
 */
export interface CertificateData {
  /** SHA-256 or SHA-512 hash of the data to be certified. */
  hash: string;
  /**
   * Optional metadata to attach to the certificate.
   * Pass a plain object and the SDK will serialize it to JSON automatically,
   * so you don't need to call `JSON.stringify` yourself.
   */
  metadata?: string | Record<string, unknown>;
  /**
   * Hashing algorithm used to produce the hash.
   * Common values: "SHA-256", "SHA-512".
   */
  algorithm?: string;
}
